crates/ra_hir/src/nameres.rs

```diff
@@ -153,28 +154,102 @@ impl ModuleScope {
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Default)]
+type ItemOrMacro = Either<PerNs<ModuleDef>, MacroDef>;
+
+trait ItemOrMacroExt {
+    fn none() -> ItemOrMacro;
+    fn take_types(self) -> Option<ModuleDef>;
+    fn is_none(&self) -> bool;
+    fn as_macros(&self) -> Option<&MacroDef>;
+    fn as_module_def(&self) -> Option<&PerNs<ModuleDef>>;
+    fn or(self, other: ItemOrMacro) -> ItemOrMacro;
+}
+
+impl ItemOrMacroExt for ItemOrMacro {
+    fn none() -> ItemOrMacro {
+        Either::Left(PerNs::none())
+    }
+
+    fn take_types(self) -> Option<ModuleDef> {
+        match self {
+            Either::Left(m) => m.take_types(),
+            _ => None,
+        }
+    }
+
+    fn is_none(&self) -> bool {
+        match self {
+            Either::Left(m) => m.is_none(),
+            Either::Right(_) => false,
+        }
+    }
+
+    fn as_macros(&self) -> Option<&MacroDef> {
+        match self {
+            Either::Right(m) => Some(m),
+            _ => None,
+        }
+    }
+
+    fn as_module_def(&self) -> Option<&PerNs<ModuleDef>> {
+        match self {
+            Either::Left(m) => Some(m),
+            _ => None,
+        }
+    }
+
+    fn or(self, other: ItemOrMacro) -> ItemOrMacro {
+        match (self, other) {
+            (Either::Left(s), Either::Left(o)) => Either::Left(s.or(o)),
+            (Either::Right(s), _) => Either::Right(s),
+            (Either::Left(s), Either::Right(o)) => {
+                if !s.is_none() {
+                    Either::Left(s)
+                } else {
+                    Either::Right(o)
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct Resolution {
     /// None for unresolved
-    pub def: PerNs<ModuleDef>,
+    def: ItemOrMacro,
```

Comment:
I am not sure it's a good idea to extend Resolution with support for macros. The problem is that `Resolution`s are stored in an `FxHashMap<Name, Resolution>`, which works for items which are guaratreed to be unique at the module scope.

A macro, however, can be defined twice inside a module:

```rust
macro_rules! foo {
  () => (fn a() {})
}
foo!();

macro_rules! foo {
  () => (fn b() {})
}
foo!();
```

So it seems like we'll need some special handling of macros anyway, so perhaps it's better to change `ModuleScope` to

```
pub struct ModuleScope {
    items: FxHashMap<Name, Resolution>,
    macro: FxHashMap<Name, MacroDef>,
}
```

this won't support macro shadowing, but it would be easier to add such support later.

Reply:
Fixed